src/main/kotlin/org/jitsi/jibri/service/impl/FileRecordingJibriService.kt

```diff
@@ -0,0 +1,174 @@
+import org.jitsi.jibri.service.JibriService
+import org.jitsi.jibri.service.JibriServiceStatus
+import org.jitsi.jibri.sink.Sink
+import org.jitsi.jibri.sink.impl.FileSink
+import org.jitsi.jibri.util.NameableThreadFactory
+import org.jitsi.jibri.util.ProcessMonitor
+import org.jitsi.jibri.util.WritableDirectory
+import org.jitsi.jibri.util.extensions.error
+import java.io.File
+import java.io.IOException
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import java.util.logging.Logger
+
+data class RecordingOptions(
+    /**
+     * The directory in which recordings should be created
+     */
+    val recordingDirectory: WritableDirectory,
+    /**
+     * The params needed to join the call
+     */
+    val callParams: CallParams,
+    /**
+     * The filesystem path to the script which should be executed when
+     *  the recording is finished.
+     */
+    val finalizeScriptPath: String
+)
+
+/**
+ * Set of metadata we'll put alongside the recording file(s)
+ */
+data class RecordingMetadata(
+    val participants: List<Map<String, Any>>
+)
+
+/**
+ * [FileRecordingJibriService] is the [JibriService] responsible for joining
+ * a web call, capturing its audio and video, and writing that audio and video
+ * to a file to be replayed later.
+ */
+class FileRecordingJibriService(private val recordingOptions: RecordingOptions) : JibriService() {
+    /**
+     * The [Logger] for this class
+     */
+    private val logger = Logger.getLogger(this::class.qualifiedName)
+    /**
+     * Used for the selenium interaction
+     */
+    private val jibriSelenium = JibriSelenium(
+        JibriSeleniumOptions(recordingOptions.callParams),
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("JibriSelenium"))
+    )
+    /**
+     * The [FfmpegCapturer] that will be used to capture media from the call and write it to a file
+     */
+    private val capturer = FfmpegCapturer()
+    /**
+     * The [Sink] this class will use to model the file on the filesystem
+     */
+    private var sink: Sink
+    /**
+     * If ffmpeg dies for some reason, we want to restart it.  This [ScheduledExecutorService]
+     * will run the process monitor in a separate thread so it can check that it's running on its own
+     */
+    private val executor: ScheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("FileRecordingJibriService"))
+    /**
+     * The handle to the scheduled process monitor task, which we use to
+     * cancel the task
+     */
+    private var processMonitorTask: ScheduledFuture<*>? = null
+
+    init {
+        sink = FileSink(recordingOptions.recordingDirectory, recordingOptions.callParams.callUrlInfo.callName)
+        jibriSelenium.addStatusHandler {
+            publishStatus(it)
+        }
+    }
+
+    override fun start(): Boolean {
```

Comment:
I wish we could avoid the duplication between `FileRecordingJibriService#start` and `StreamingJibriService#start`. They are very similar (and not trivial)

Reply:
Same goes for `stop`. They both have a `capturer` and a `processMonitorTask`. Perhaps we can move these to super?

Reply:
yeah totally agree.  i tried to combine them earlier on but it was a bit awkward and because of some of the differences still didn't feel great.  i want to take another look at it now; especially with adding the sipgw stuff there is more duplication so hoping can find a way to reuse some code.